src/common/nlmeans_core.c

```diff
@@ -0,0 +1,922 @@
+#include "develop/tiling.h"
+#include "iop/iop_api.h"
+#include "common/nlmeans_core.h"
+#include <stdbool.h>
+#include <stdlib.h>
+
+#if defined(__SSE2__)
+#include <xmmintrin.h>
+#endif
+
+// to avoid accumulation of rounding errors, we should do a full recomputation of the patch differences
+//   every so many rows of the image.  We'll also use that interval as the target maximum chunk size for
+//   parallelization
+// lower values reduce the accumulation of rounding errors at the cost of slightly more computation, but
+//  also improve cache sharing between threads (on the boundaries of slices), resulting in better scaling
+//  because memory bandwidth is less of a limitation.  The value of 20 is a good tradeoff between
+//  single-threaded performance and highly-threaded performance; 12 was found to be best on a 32-core
+//  third-gen Threadripper.
+#define SLICE_HEIGHT 20
+
+// number of intermediate buffers used by OpenCL code path.  If you change this, you must also change
+//   the definition in src/iop/nlmeans.c and src/iop/denoiseprofile.c
+#define NUM_BUCKETS 4
+
+// a structure to collect together the items which define the location of a patch relative to the pixel
+//  being denoised
+struct patch_t
+{
+  short rows;		// number of rows difference
+  short cols;		// number of columns difference
+  int offset;	  	// array distance between corresponding pixels
+};
+typedef struct patch_t patch_t;
+
+// some shorthand to make code more legible
+// if we have OpenMP simd enabled, declare a vectorizable for loop;
+// otherwise, just leave it a plain for()
+#if defined(_OPENMP) && defined(OPENMP_SIMD_)
+#define SIMD_FOR \
+  _Pragma("omp simd") \
+  for
+#else
+#define SIMD_FOR for
+#endif
+
+
+typedef union floatint_t
+{
+  float f;
+  uint32_t i;
+} floatint_t;
+
+static inline float gh(const float f)
+{
+#if 0
+  return exp2f(-f);
+#else
+  // fast integer-hack version of the above
+  const int i1 = 0x3f800000; // 2^0
+  const int i2 = 0x3f000000; // 2^-1
+  const int k0 = i1 + (int)(f * (i2 - i1));
+  floatint_t k;
+  k.i = k0 >= 0x800000 ? k0 : 0;
+  return k.f;
+#endif
+}
+
+static inline int sign(int a)
+{
+  return (a > 0) - (a < 0);
+}
+
+// map the basic row/column offset into a possible much larger offset based on a user parameter
+static int scatter(float scale, int scattering, int index1, int index2)
+{
+  // this formula is designed to
+  //  - produce an identity mapping when scattering = 0
+  //  - avoiding duplicate patches provided that 0 <= scattering <= 1
+  //  - avoiding grid artifacts by trying to take patches on various rows and columns
+  const int abs_i1 = abs(index1);
+  const int abs_i2 = abs(index2);
+  return scale * ((abs_i1 * abs_i1 * abs_i1 + 7.0 * abs_i1 * sqrt(abs_i2)) * sign(index1) * scattering / 6.0 + index1);
+}
+
+// allocate and fill an array of patch definitions
+static struct patch_t*
+define_patches(const dt_nlmeans_param_t *const params, const int stride, int *num_patches, int *max_shift)
```

Comment:
If I understand correctly, this defines an array with all the patches positions. Maybe there could be some performance speedup if we sort this array according to the offset (scattering was designed to produce a mess out of the normal offsets, and as such it does not give an ordered output from an ordered input), to avoid some cache misses?

Reply:
You understand correctly.  And once the refactored CL code path is validated as correct, we can give the nlmeans iop the scattering feature just by wiring up the GUI (the power of sharing code instead of making wholesale copies as I'm seeing so frequently in the codebase).

Sorting the array is certainly a possibility that I haven't looked into yet.  It did occur to me that we could omit the center pixel (offset 0,0) from the patches array and initialize the output buffer with a copy of the input buffer for about an 0.4% speedup (you save one patch computation out of 225).

What might help more than sorting strictly by offset is sorting such that patches which will often occupy the same or adjacent cache lines get grouped together.  That might be a bit tricky to implement, and would be moot for very small scattering values.

I'll give the sorting a try once I finish my current work speeding up wavelets denoise.  Only 3% single threaded so far, but 25% with lots of threads thanks to integrating some formerly single-threaded computation as part of existing parallelized code.